### PR TITLE
Waku post

### PR DESCRIPTION
**Pull request by oskarth**
Initial draft

### CODE REVIEW

File: _posts/2020-02-10-waku-current-state.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+summary: A research log. Waku current state.
+image: /assets/img/TODOFIXME.png
+discuss: https://forum.vac.dev/t/TODOFIXME
+---
+
+TODO: Picture for bottleneck or Waku
+
+# Where we are at with Waku?
+
+Waku is our fork of Whisper where we address the shortcomings of Whisper in an iterative manner. We've seen in previous post that Whisper doesn't scale, and why. Here's an update on where we are at with Waku since that [last post](https://vac.dev/fixing-whisper-with-waku).
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
Waku is our fork of Whisper where we address the shortcomings of Whisper in an iterative manner. We've seen a in [previous post](https://vac.dev/fixing-whisper-with-waku) that Whisper doesn't scale, and why. Here's an update on where we are at with Waku since then.
```

File: _posts/2020-02-10-waku-current-state.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+- Waku envelope data field, currently in [version 0.1.0](https://specs.vac.dev/waku/envelope-data-format.html)
+- Waku mailserver, currently in [version 0.2.0](https://specs.vac.dev/waku/mailserver.html)
+
+TODO: Update Waku spec to 0.3
+
+**Clients:**
+
+There are currently two clients that implement Waku, these are [Nimbus](https://github.com/status-im/nimbus/tree/master/waku) in Nim and [status-go](https://github.com/status-im/status-go) in Go.
+
+At the time of this writing the Nimbus client implements the spec fully, but lacks mail server and mail client capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
At the time of writing this, the Nimbus client implements the spec fully, but lacks mail server and mail client capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
```

File: _posts/2020-02-10-waku-current-state.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+For more details, see [implementation matrix](https://specs.vac.dev/waku/waku.html#appendix-b-implementation-notes).
+
+In terms of end user applications, work is currently in progress to integrate it into the [Status core app](https://github.com/status-im/status-react/pull/9949) and is expected to be released in their upcoming 1.1 release.
+
+TODO: Fact check this with Adam and Kim
+
+**Simulation:**
+
+We've got a [simulation](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol) in the Nimbus client that verifies - or rather, fails to falsify - the scalability model posted in an [earlier post](https://vac.dev/fixing-whisper-with-waku). More on this below.
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
We've got a [simulation](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol) in the Nimbus client that verifies - or rather, fails to falsify - the scalability model described in an [earlier post](https://vac.dev/fixing-whisper-with-waku). More on this below.
```

File: _posts/2020-02-10-waku-current-state.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+Things to note:
+- Whisper light node with 1% false positive gets 1% of total traffic
+- Waku light node gets ~1000x less envelopes than Whisper light node
+- Full mesh results in a lot more duplicate messages, expect for Waku light node
+
+Run the simulation yourself [here](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol). The parameters are configurable, and it is integrated with Prometheus and Grafana.
+
+## Difference between Waku and Whisper
+
+Summary of main differences between Waku v0 spec and Whisper v6, as described in EIP-627:
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
Summary of main differences between Waku v0 spec and Whisper v6, as described in [EIP-627](https://eips.ethereum.org/EIPS/eip-627):
```

File: _posts/2020-02-10-waku-current-state.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+## Simulation
+
+The ultimate test is real-world usage. Until then, we have a simulation. Here's some data on this.
+
+![](assets/img/waku_simulation.jpeg)
+
+Here we outline two network topologies, Star and full mesh, with 6 randomly connected nodes and 2 X, sending 10000 random envelopes over random topics, including invalid envelopes.
+
+For light node, bloom filter is set to 1% false positive. It shows the number of valid and invalid envelopes received for the different nodes.
```
Comment (kdeme) [inline _posts/2020-02-10-waku-current-state.md]: With the settings in the quicksim the false positive probability is almost 10% (p=0.087). 
Light node and Waku light node subscribe to 100 randomly selected topics (that makes n=100, k=3, m=512 for the bloom filter). 

Next, one of the full nodes sends 1 envelope over 1 of the 100 topics that the two light nodes subscribed on. This is to test that the subscription works, else the 10k envelopes will of course not be received (come to think of it, this should be done for all 100 topics to be a proper check).
Then, that same full node sends 10000 envelopes over random topics.

This explains why you see 10001 for the full nodes, and 1 for the Waku light node (as in fact, with just 10000 envelopes to random topics, the chance is very low that one of the Waku light node topics gets "hit").

File: _posts/2020-02-10-waku-current-state.md (hunk 2 above)
Comment (kdeme) [inline _posts/2020-02-10-waku-current-state.md]: Not sure if this detail needs to be added, but the nim implementation also lacks rate limiting and confirmations. But these are optional so in theory it still complies to spec.

File: _posts/2020-02-10-waku-current-state.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+**Simulation:**
+
+We've got a [simulation](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol) in the Nimbus client that verifies - or rather, fails to falsify - the scalability model posted in an [earlier post](https://vac.dev/fixing-whisper-with-waku). More on this below.
+
+## How many users does Waku support?
+
+This is our current understanding of how many users a network running the Waku protocol can support. Specifically in the context of the Status chat app, since that's the most immediate consumer of Waku. It should generalize fairly well to most deployments, but YMMW.
+
+**tldr (for Status app):**
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
**tl;dr (for Status app):**
```

File: _posts/2020-02-10-waku-current-state.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+- beta: 100 DAU
+- v1: 1k DAU
+- v1.1 (waku only): 10k DAU (up to x10 with deployment hotfixes)
+- v1.2 (waku+dns): 100k DAU (can optionally be folded into v1.1)
+
+*Assuming 10 concurrent users = 100 DAU. Estimate uncertainty increases for each order of magnitude until real-world data is observed.*
+
+As far as we know right now, there is one immediate bottleneck, a very likely bottleneck after that, and a third bottleneck that is still conjecture but not unlikely to appear. These are:
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
As far as we know right now, these are the bottlenecks we have:
```

File: _posts/2020-02-10-waku-current-state.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+- Bottleneck 1 - Receive bandwidth for end user clients (aka ‘Fixing Whisper with Waku’)
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
- Immediate bottleneck - Receive bandwidth for end user clients (aka ‘Fixing Whisper with Waku’)
```

File: _posts/2020-02-10-waku-current-state.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+- Bottleneck 2 - Nodes and cluster capacity (aka ‘DNS based node discovery’)
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
- Very likely bottleneck - Nodes and cluster capacity (aka ‘DNS based node discovery’)
```

File: _posts/2020-02-10-waku-current-state.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+- Bottleneck 3 - Full node traffic (aka ‘the routing / partition problem’)
```
Comment (decanus) [inline _posts/2020-02-10-waku-current-state.md]:
```suggestion
- Conjecture but not unlikely to appear- Full node traffic (aka ‘the routing / partition problem’)
```

File: _posts/2020-02-10-waku-current-state.md (hunk 2 above)
Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]:
```suggestion
At the time of writing the Nimbus client implements the spec fully, but lacks mail server and mail client capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
```
https://english.stackexchange.com/questions/199381/is-at-the-time-of-writing-correct/199392

File: _posts/2020-02-10-waku-current-state.md (hunk 5 above)
Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]: Ah should've looked more carefully, thanks for the correction and added detail :)

File: _posts/2020-02-10-waku-current-state.md (hunk 7 above)
Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]: Hm, there's only one bottleneck at a time though? They don't appear all at once

Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]: I suppose you edit this below

File: _posts/2020-02-10-waku-current-state.md (hunk 2 above)
Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]: added

File: _posts/2020-02-10-waku-current-state.md (hunk 5 above)
Comment (oskarth) [reply _posts/2020-02-10-waku-current-state.md]: fixed

File: _posts/2020-02-10-waku-current-state.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+- Waku (main spec), currently in [version 0.3.0](https://specs.vac.dev/waku/waku.html)
+- Waku envelope data field, currently in [version 0.1.0](https://specs.vac.dev/waku/envelope-data-format.html)
+- Waku mailserver, currently in [version 0.2.0](https://specs.vac.dev/waku/mailserver.html)
+
+**Clients:**
+
+There are currently two clients that implement Waku, these are [Nimbus](https://github.com/status-im/nimbus/tree/master/waku) in Nim and [status-go](https://github.com/status-im/status-go) in Go.
+
+At the time of writing the Nimbus client implements the spec fully, but lacks mail server/client, rate limiting and confirmations capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
```
Comment (oskarth) [inline _posts/2020-02-10-waku-current-state.md]: @adambabik please confirm this is accurate, ideally with link to whats left

File: _posts/2020-02-14-where-are-we-at-with-waku.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+- Waku (main spec), currently in [version 0.3.0](https://specs.vac.dev/waku/waku.html)
+- Waku envelope data field, currently in [version 0.1.0](https://specs.vac.dev/waku/envelope-data-format.html)
+- Waku mailserver, currently in [version 0.2.0](https://specs.vac.dev/waku/mailserver.html)
+
+**Clients:**
+
+There are currently two clients that implement Waku, these are [Nimbus](https://github.com/status-im/nimbus/tree/master/waku) in Nim and [status-go](https://github.com/status-im/status-go) in Go.
+
+At the time of writing the Nimbus client implements the spec fully, but lacks mail server/client, rate limiting and confirmations capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
```
Comment (adambabik) [inline _posts/2020-02-14-where-are-we-at-with-waku.md]: status-go has partial rate-limiting implementation. There is still work to be done yet in terms of broadcasting it and enforcing.

File: _posts/2020-02-14-where-are-we-at-with-waku.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+Run the simulation yourself [here](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol). The parameters are configurable, and it is integrated with Prometheus and Grafana.
+
+## Difference between Waku and Whisper
+
+Summary of main differences between Waku v0 spec and Whisper v6, as described in [EIP-627](https://eips.ethereum.org/EIPS/eip-627):
+
+- Handshake/Status message not compatible with shh/6 nodes; specifying options as association list
+- Include topic-interest in Status handshake
+- Upgradability policy
+- `topic-interest` packet code
```
Comment (adambabik) [inline _posts/2020-02-14-where-are-we-at-with-waku.md]: There has been some developments around it to remove it and replace with a packet to exchange `statusOptions` like in the handshake.

File: _posts/2020-02-14-where-are-we-at-with-waku.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+**Clients:**
+
+There are currently two clients that implement Waku, these are [Nimbus](https://github.com/status-im/nimbus/tree/master/waku) in Nim and [status-go](https://github.com/status-im/status-go) in Go.
+
+At the time of writing the Nimbus client implements the spec fully, but lacks mail server/client, rate limiting and confirmations capability. The status-go client implements everything except bridging mode, which is currently a work in progress.
+
+For more details, see [implementation matrix](https://specs.vac.dev/waku/waku.html#appendix-b-implementation-notes).
+
+In terms of end user applications, work is currently in progress to integrate it into the [Status core app](https://github.com/status-im/status-react/pull/9949) and is expected to be released in their upcoming 1.1 release.
```
Comment (decanus) [inline _posts/2020-02-14-where-are-we-at-with-waku.md]: any link to show its coming in 1.1?

File: _posts/2020-02-14-where-are-we-at-with-waku.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+TODO: Fact check this with Adam and Kim
+
+**Simulation:**
+
+We've got a [simulation](https://github.com/status-im/nimbus/tree/master/waku#testing-waku-protocol) in the Nimbus client that verifies - or rather, fails to falsify - the scalability model described in an [earlier post](https://vac.dev/fixing-whisper-with-waku). More on this below.
+
+## How many users does Waku support?
+
+This is our current understanding of how many users a network running the Waku protocol can support. Specifically in the context of the Status chat app, since that's the most immediate consumer of Waku. It should generalize fairly well to most deployments, but YMMW.
```
Comment (decanus) [inline _posts/2020-02-14-where-are-we-at-with-waku.md]: what is YMMW?

Comment (oskarth) [reply _posts/2020-02-14-where-are-we-at-with-waku.md]: your milage might vary (typo)

removed

File: _posts/2020-02-14-where-are-we-at-with-waku.md (hunk 14 above)
Comment (oskarth) [reply _posts/2020-02-14-where-are-we-at-with-waku.md]: added link to roadmap, no better link right now

**Comment by oskarth** [comment]
Addressed comments. Removing you as reviewer @decanus as I fixed those things and want to push it live to preview. We can always amend in follow up PR
